src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs

```diff
@@ -354,6 +383,138 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
+
+        /// <summary>
+        /// Update the resource wrapper if it's have the "Id" value.
+        /// </summary>
+        /// <param name="resourceWrapper">The resource wrapper.</param>
+        /// <param name="readContext">The read context.</param>
+        /// <returns>The resource wrapper.</returns>
+        private ODataResourceWrapper UpdateResourceWrapper(ODataResourceWrapper resourceWrapper, ODataDeserializerContext readContext)
+        {
+            Contract.Assert(readContext != null);
+
+            if (resourceWrapper?.Resource?.Id == null)
+            {
+                return resourceWrapper;
+            }
+
+            IEnumerable<ODataProperty> keys = CreateKeyProperties(resourceWrapper.Resource.Id, readContext);
+            if (keys == null)
+            {
+                return resourceWrapper;
+            }
+
+            if (resourceWrapper.Resource.Properties == null)
+            {
+                resourceWrapper.Resource.Properties = keys;
+            }
+            else
+            {
+                IDictionary<string, ODataProperty> newPropertiesDic = resourceWrapper.Resource.Properties.ToDictionary(p => p.Name, p => p);
+                foreach (ODataProperty key in keys)
+                {
+                    if (!newPropertiesDic.ContainsKey(key.Name))
+                    {
+                        newPropertiesDic[key.Name] = key;
+                    }
+                }
+
+                resourceWrapper.Resource.Properties = newPropertiesDic.Values;
+            }
+
+            return resourceWrapper;
+        }
+
+        /// <summary>
+        /// Do uri parsing to get the key values.
+        /// </summary>
+        /// <param name="id">The key Id.</param>
+        /// <param name="readContext">The reader context.</param>
+        /// <returns>The key properties.</returns>
+        private static IList<ODataProperty> CreateKeyProperties(Uri id, ODataDeserializerContext readContext)
+        {
+            Contract.Assert(id != null);
+            Contract.Assert(readContext != null);
+            IList<ODataProperty> properties = new List<ODataProperty>();
+            if (readContext.Request == null)
+            {
+                return properties;
+            }
+
+            try
+            {
+                IODataPathHandler pathHandler = readContext.InternalRequest.PathHandler;
+                IWebApiRequestMessage internalRequest = readContext.InternalRequest;
+                IWebApiUrlHelper urlHelper = readContext.InternalUrlHelper;    
+                string serviceRoot = urlHelper.CreateODataLink(
+                    internalRequest.Context.RouteName,
+                    internalRequest.PathHandler,
+                    new List<ODataPathSegment>());
+
+                ODataPath odataPath = pathHandler.Parse(serviceRoot, id.OriginalString, internalRequest.RequestContainer);
+                IList<KeySegment> keySegments = odataPath.Segments.OfType<KeySegment>().ToList();
```

Comment:
I think we only expect to have one key segment right? If we have more than one key segments, then that means the url has navigation properties and not all the keys belong to the same resource, e.g. `http://localhost/odata/Books(1)/Authors(2)`.

I assume we don't expect such a scenario to occur in `@odata.bind` or `@odata.id`. Or am I wrong?

And in such cases, it wouldn't make sense to add both the keys of `Book` and `Author` to the same resource. So maybe you should pick the first key segment instead of going through all of them. Or just pick the last key segment if it makes sense to have a url with multiple key segments.

Reply:
@habbes Question, maybe unrelated to your point here, how does OData handle a composite key?

Reply:
@gathogojr in the case of a composite key, the `KeySegment` will have multiple keys. `keySegment.Keys` is a collection of key-value pairs, mapping the key property names to their values.

Reply:
@habbes Composite key question out of the way, you raise an important question here. If something like `/Books(1)/Authors(2)` is allowed, then we'd probably need to target the last key segment. We might need some guidance here @mikepizzo